streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskExecutor.java

```diff
@@ -175,14 +175,14 @@ void commitOffsetsOrTransaction(final Map<Task, Map<TopicPartition, OffsetAndMet
 
         final Set<TaskId> corruptedTasks = new HashSet<>();
 
-        if (!offsetsPerTask.isEmpty()) {
-            if (executionMetadata.processingMode() == EXACTLY_ONCE_ALPHA) {
-                for (final Map.Entry<Task, Map<TopicPartition, OffsetAndMetadata>> taskToCommit : offsetsPerTask.entrySet()) {
-                    final Task task = taskToCommit.getKey();
+        if (executionMetadata.processingMode() == EXACTLY_ONCE_ALPHA) {
+            for (final Task task : taskManager.activeTaskIterable()) {
+                final Map<TopicPartition, OffsetAndMetadata> offsetsToCommit = offsetsPerTask.get(task);
+                if (offsetsToCommit != null || taskManager.streamsProducerForTask(task.id()).transactionInFlight()) {
                     try {
                         taskManager.streamsProducerForTask(task.id())
-                            .commitTransaction(taskToCommit.getValue(), taskManager.mainConsumer().groupMetadata());
-                        updateTaskCommitMetadata(taskToCommit.getValue());
+                            .commitTransaction(offsetsToCommit, taskManager.consumerGroupMetadata());
```

Comment:
Wouldn't this lead to an NPE if there are no offsets to commit but a transaction is in flight?
See https://github.com/apache/kafka/blob/7387a1162533c52c83fb8899a31c505c42e0fe10/clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java#L752

Reply:
ah, yes, we should pass in an empty map not `null`